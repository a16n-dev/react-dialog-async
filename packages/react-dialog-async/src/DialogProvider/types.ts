import type { PropsWithChildren } from 'react';

export interface DialogProviderProps extends PropsWithChildren {
  /**
   * The default delay in milliseconds to wait before unmounting a dialog after it's closed.
   */
  defaultUnmountDelayInMs?: number;
}
