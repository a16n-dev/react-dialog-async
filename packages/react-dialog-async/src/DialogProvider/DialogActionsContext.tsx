import { createContext } from 'react';
import type { AsyncDialogComponent } from '../types.js';

export interface DialogActionsContextValue {
  show: (
    dialogId: string,
    hash: number,
    dialog: AsyncDialogComponent<any, any>,
    data: unknown,
    unmountDelay?: number,
  ) => Promise<any>;
  hide: (dialogId: string, data?: any) => void;
  updateData: (dialogId: string, data: unknown) => void;
}

export const DialogActionsContext =
  createContext<DialogActionsContextValue | null>(null);
