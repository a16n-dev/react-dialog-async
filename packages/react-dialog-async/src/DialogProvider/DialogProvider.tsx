import React, { useContext, useEffect, useMemo, useRef } from 'react';
import { PropsWithChildren, useCallback, useState } from 'react';
import DialogContext, { dialogContextState } from '../DialogContext';
import { DialogComponent } from '../types';
import DialogStateContext, { dialogsStateData } from '../DialogStateContext';
import useRenderDialogs from '../useRenderDialogs';

interface DialogProviderProps extends PropsWithChildren {
  /**
   * The default delay in milliseconds to wait before unmounting a dialog after it's closed.
   */
  defaultUnmountDelayInMs?: number;
  /**
   * If provided, this function will be called when useDialogLazy is mounted.
   * This lets you call the preload function at some point before the lazy component is required
   * Consult the docs for some sensible functions to provide here for different environments
   */
  lazyLoaderFn?: (preload: () => Promise<void>) => Promise<void>;
}

const DialogProvider = ({
  defaultUnmountDelayInMs,
  lazyLoaderFn,
  children,
}: DialogProviderProps) => {
  // This ref tracks timers for unmount dialogs after they're closed
  const unmountDelayTimeoutRefs = useRef<{ [key: string]: any }>({});

  const [dialogState, setDialogState] = useState<dialogsStateData>({});
  const [usingOutlet, setUsingOutlet] = useState(false);

  /**
   * Force closes the dialog with the given id.
   */
  const hide = useCallback((id: string) => {
    setDialogState((state) => {
      if (!state[id]) return state;

      if (!state[id].open) return state; // don't do anything if the dialog is already closed

      state[id].resolve?.(undefined);

      if (state[id].unmountDelay) {
        if (unmountDelayTimeoutRefs.current[id] !== undefined) {
          clearTimeout(unmountDelayTimeoutRefs.current[id]);
        }

        // start the delay
        unmountDelayTimeoutRefs.current[id] = setTimeout(() => {
          setDialogState((state) => removeKey(state, id));
        }, state[id].unmountDelay);

        return {
          ...state,
          [id]: {
            open: false,
            dialog: state[id].dialog,
            data: state[id].data,
          },
        };
      }
      return removeKey(state, id);
    });
  }, []);

  const show = useCallback(
    (
      id: string,
      hash: number,
      dialog: DialogComponent<any, any>,
      data: unknown,
      unmountDelay?: number,
    ): Promise<unknown> => {
      return new Promise((resolve) => {
        if (unmountDelayTimeoutRefs.current[id] !== undefined) {
          clearTimeout(unmountDelayTimeoutRefs.current[id]);
        }

        const resolveFn = (value: any) => {
          resolve?.(value);
          hide(id);
        };

        setDialogState((state) => {
          if (state[id]?.open) {
            resolve(undefined);
            return state;
          }

          return {
            ...state,
            [id]: {
              dialog: dialog,
              open: true,
              hash,
              data,
              resolve: resolveFn,
              unmountDelay: unmountDelay ?? defaultUnmountDelayInMs,
            },
          };
        });
      });
    },
    [hide, defaultUnmountDelayInMs],
  );

  /**
   * Updates the data of the given dialog
   */
  const updateData = useCallback((id: string, data: unknown): void => {
    setDialogState((state) => {
      if (state[id]) {
        return {
          ...state,
          [id]: { ...state[id], data },
        };
      }
      return state;
    });
  }, []);

  useEffect(() => {
    return () => {
      Object.values(unmountDelayTimeoutRefs.current).forEach(clearTimeout);
    };
  }, []);

  /**
   * To prevent unnecessary re-renders, be careful to ensure that this state has
   * no transitive dependencies to `dialogState`
   */
  const ctx: dialogContextState = useMemo(
    () => ({
      show,
      hide,
      updateData,
      lazyLoaderFn,
    }),
    [show, hide, updateData],
  );

  return (
    <DialogStateContext.Provider
      value={{ dialogs: dialogState, setIsUsingOutlet: setUsingOutlet }}
    >
      <DialogContext.Provider value={ctx}>
        {children}
        {!usingOutlet && <InternalDialogOutlet />}
      </DialogContext.Provider>
    </DialogStateContext.Provider>
  );
};

export default DialogProvider;

function removeKey<
  S extends string | number | symbol,
  T extends Record<S, unknown>,
>(data: T, key: S): Omit<T, S> {
  const { [key]: _, ...rest } = data;
  return rest;
}

const InternalDialogOutlet = () => {
  const dialogState = useContext(DialogStateContext);

  if (!dialogState) {
    throw new Error(
      'Dialog context not found. You likely forgot to wrap your app in a <DialogProvider/> (https://react-dialog-async.a16n.dev/installation)',
    );
  }

  const dialogComponents = useRenderDialogs(dialogState.dialogs);

  if (process.env.NODE_ENV !== 'production' && dialogComponents.length > 0) {
    console.warn(
      'Rendering a dialog without a <DialogOutlet/>. Please include a <DialogOutlet/> as a child of <DialogProvider/> to ensure dialogs are rendered within the correct contexts - this will be required in the next major version of react-dialog-async. See https://react-dialog-async.a16n.dev/API/dialog-outlet for more details. This warning is only present in development',
    );
  }

  return <>{dialogComponents}</>;
};
