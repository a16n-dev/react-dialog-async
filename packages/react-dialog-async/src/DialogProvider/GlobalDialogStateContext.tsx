import { createContext } from 'react';
import type { AsyncDialogComponent } from '../types.js';

export type dialogsStateData = Record<
  string,
  {
    dialog: AsyncDialogComponent<any, any>;
    hash: number;
    data: unknown;
    open: boolean;
    resolve?: (value?: unknown) => void;
    unmountDelay?: number;
  }
>;

export type GlobalDialogStateContextValue = {
  setIsUsingOutlet: (value: boolean) => void;
  dialogs: dialogsStateData;
};

export const GlobalDialogStateContext =
  createContext<GlobalDialogStateContextValue | null>(null);
