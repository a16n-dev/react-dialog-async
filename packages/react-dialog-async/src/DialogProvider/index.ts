export { default as DialogProvider } from './DialogProvider';
export { type DialogProviderProps } from './types';
