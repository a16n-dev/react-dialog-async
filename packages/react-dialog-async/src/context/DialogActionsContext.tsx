import { createContext } from 'react';
import type { AsyncDialogComponent } from '../types.js';

export interface DialogActionsContextValue {
  show: (
    dialogId: string,
    hash: number,
    dialog: AsyncDialogComponent<any, any>,
    data: unknown,
    unmountDelay?: number,
  ) => Promise<any>;
  hide: (dialogId: string, data?: any) => void;
  updateData: (dialogId: string, data: unknown) => void;
  lazyLoaderFn?: (loaderFn: () => Promise<void>) => Promise<void>;
}

export const DialogActionsContext =
  createContext<DialogActionsContextValue | null>(null);
