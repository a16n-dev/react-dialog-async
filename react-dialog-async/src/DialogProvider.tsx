import React, { PropsWithChildren, useRef } from 'react';
import DialogContext, { dialogContextState } from './DialogContext';
import { DialogComponent } from './types';

interface dialogStateItem {
  component: DialogComponent<unknown, unknown>;
  isOpen: boolean;
  data?: unknown;
  resolve?: (value: unknown) => void;
}

const DialogProvider = ({ children }: PropsWithChildren<{}>) => {
  const [dialogs, setDialogs] = React.useState<{ [key: string]: dialogStateItem }>({});
  const id = useRef(0);

  const ctx: dialogContextState = {
    /**
     * Registers a new dialog component and returns the assigned id of that dialog
     * @param component The component to use
     * @returns the id of the dialog
     */
    register: (component: DialogComponent<unknown, unknown>): string => {
      const dialog: dialogStateItem = {
        component,
        isOpen: false,
      };

      const componentId = id.current;
      id.current++;
      
      setDialogs((d) => ({ ...d, [componentId]: dialog }));
      return String(componentId);
    },

    /**
     * Removes a registerd dialog given its id
     * @param dialogId the id of the dialog
     */
    unregister: (dialogId: string): void => {
      setDialogs(({ [dialogId]: _, ...d }) => d);
    },
    /**
     * Show the dialog with the given id, and optionally pass in some data. If the dialog is already open, this will just update the data
     * @param dialogId the id of the dialog
     * @param data data to pass to the dialog component
     */
    show: (dialogId: string, data: unknown): Promise<unknown> => {
      return new Promise((resolve) => {
        setDialogs((d) => ({ ...d, [dialogId]: { ...d[dialogId], data, isOpen: true, resolve } }));
      });
    },
    /**
     * Force closes the dialog with the given id
     * @param dialogId the id of the dialog
     */
    hide: (dialogId: string): void => {
      setDialogs((d) => ({ ...d, [dialogId]: { ...d[dialogId], isOpen: false } }));
    },
    
    /**
     * Updates the data of the given dialog
     * @param dialogId the id of the dialog to update
     * @param data the new data to pass to the dialog
     */
    updateData: (dialogId: string, data: unknown): void => {
      setDialogs((d) => ({ ...d, [dialogId]: { ...d[dialogId], data } }));
    },
  };

  return (
    <>
      <DialogContext.Provider value={ctx}>
        {children}
        {Object.entries(dialogs).map(([id, { component: C, isOpen, data, resolve }]) => {
          if (resolve) {
            const onClose = (value: unknown) => {
              ctx.hide(id);
              resolve(value);
            };
            return <C key={id} open={isOpen} handleClose={onClose} data={data} />;
          } else {
            return; 
          }
        })}
      </DialogContext.Provider>
    </>
  );
};

export default DialogProvider;
