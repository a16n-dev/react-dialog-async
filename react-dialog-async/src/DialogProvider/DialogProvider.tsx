import { PropsWithChildren, useMemo, useRef, useState } from 'react';
import DialogContext, { dialogContextState } from '../DialogContext';
import { DialogComponent } from '../types';

interface dialogStateItem {
  /**
   * The dialog component to render
   */
  component: DialogComponent<unknown, unknown>;
  /**
   * Represents if the dialog is currently visible
   */
  isOpen: boolean;
  /**
   * The data that should be passed to the dialog
   */
  data?: unknown;
  /**
   * The function that should be called when the dialog is closed
   */
  resolve?: (value: unknown) => void;
}

const DialogProvider = ({ children }: PropsWithChildren) => {
  const [dialogs, setDialogs] = useState<{ [key: string]: dialogStateItem }>(
    {},
  );
  const id = useRef(0);

  const dialogComponents = useMemo(() => {
    return Object.entries(dialogs).map(([id, dialog]) => {
      const resolve = dialog.resolve;

      if (!resolve) return;
      const Component = dialog.component;
      const onClose = (value: unknown) => {
        ctx.hide(id);
        resolve(value);
      };
      return (
        <Component
          key={id}
          open={dialog.isOpen}
          handleClose={onClose}
          data={dialog.data}
        />
      );
    });
  }, [dialogs]);

  /**
   * Registers a new dialog component and returns the assigned id of that dialog
   * @param component The component to use
   * @returns the id of the dialog
   */
  const register = (component: DialogComponent<unknown, unknown>): string => {
    const dialog: dialogStateItem = {
      component,
      isOpen: false,
    };

    const componentId = id.current;
    id.current++;

    setDialogs((d) => ({ ...d, [componentId]: dialog }));
    return String(componentId);
  };

  /**
   * Removes a registered dialog given its id
   * @param dialogId the id of the dialog
   */
  const unregister = (dialogId: string): void => {
    setDialogs(({ [dialogId]: _, ...d }) => d);
  };

  /**
   * Show the dialog with the given id, and optionally pass in some data. If the dialog is already open, this will just update the data
   * @param dialogId the id of the dialog
   * @param data data to pass to the dialog component
   */
  const show = (dialogId: string, data: unknown): Promise<unknown> => {
    return new Promise((resolve) => {
      setDialogs((d) => ({
        ...d,
        [dialogId]: { ...d[dialogId], data, isOpen: true, resolve },
      }));
    });
  };

  /**
   * Force closes the dialog with the given id
   * @param dialogId the id of the dialog
   */
  const hide = (dialogId: string): void => {
    setDialogs((d) => ({
      ...d,
      [dialogId]: { ...d[dialogId], isOpen: false },
    }));
  };

  /**
   * Updates the data of the given dialog
   * @param dialogId the id of the dialog to update
   * @param data the new data to pass to the dialog
   */
  const updateData = (dialogId: string, data: unknown): void => {
    setDialogs((d) => ({ ...d, [dialogId]: { ...d[dialogId], data } }));
  };

  const ctx: dialogContextState = {
    register,
    unregister,
    show,
    hide,
    updateData,
  };

  return (
    <DialogContext.Provider value={ctx}>
      {children}
      {dialogComponents}
    </DialogContext.Provider>
  );
};

export default DialogProvider;
