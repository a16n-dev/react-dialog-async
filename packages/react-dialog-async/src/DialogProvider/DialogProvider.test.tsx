import { expect, test } from 'vitest';
import { render, screen } from '@testing-library/react';
import { useContext, useEffect, useId } from 'react';
import type { AsyncDialogProps } from '../types.js';
import { DialogActionsContext } from '../context/DialogActionsContext.js';
import { DialogProvider } from './DialogProvider.js';

function valueNotNull<T>(v: T | null): asserts v {
  if (!v) {
    throw new Error(
      'Dialog context not found. You likely forgot to wrap your app in a <DialogProvider/> (https://react-dialog-async.a16n.dev/installation)',
    );
  }
}

const TestDialog = ({ isOpen }: AsyncDialogProps) => {
  if (!isOpen) return null;

  return <div>{'Hello World!'}</div>;
};

test('renders without error', () => {
  const result = render(<DialogProvider />);

  expect(result.asFragment()).toMatchSnapshot();
});

test('renders children', () => {
  render(<DialogProvider>Hello world!</DialogProvider>);

  expect(screen.findByText('Hello world!')).toBeDefined();
});

test('calling show() mounts the dialog in the DOM', () => {
  const InnerComponent = () => {
    const ctx = useContext(DialogActionsContext);
    valueNotNull(ctx);
    const id = useId();

    useEffect(() => {
      ctx.show(id, 0, TestDialog, {});
    }, [id]);

    return null;
  };
  const result = render(
    <DialogProvider>
      <InnerComponent />
    </DialogProvider>,
  );

  expect(result.asFragment()).toMatchSnapshot();
  expect(screen.getByText('Hello World!')).toBeDefined();
});

test('calling show() followed by hide() unmounts the dialog in the DOM', () => {
  const InnerComponent = () => {
    const ctx = useContext(DialogActionsContext);
    valueNotNull(ctx);
    const id = useId();

    useEffect(() => {
      ctx.show(id, 0, TestDialog, {});

      ctx.hide(id);
    }, [id]);

    return null;
  };
  const result = render(
    <DialogProvider>
      <InnerComponent />
    </DialogProvider>,
  );

  expect(result.asFragment()).toMatchSnapshot();
  expect(screen.queryByText('Hello World!')).toBeNull();
});
