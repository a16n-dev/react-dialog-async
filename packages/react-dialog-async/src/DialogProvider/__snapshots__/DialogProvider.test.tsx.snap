// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`calling register() does not mount the dialog 1`] = `<DocumentFragment />`;

exports[`calling register() followed by show() followed by hide() unmounts the dialog in the DOM 1`] = `<DocumentFragment />`;

exports[`calling register() followed by show() mounts the dialog in the DOM 1`] = `
<DocumentFragment>
  <div>
    Hello World!
  </div>
</DocumentFragment>
`;

exports[`calling register() twice on a dialog a key results in only 1 dialog being mounted 1`] = `
<DocumentFragment>
  <div>
    Hello World!
  </div>
</DocumentFragment>
`;

exports[`calling register() twice on a dialog with no key results in two dialogs being mounted 1`] = `
<DocumentFragment>
  <div>
    Hello World!
  </div>
  <div>
    Hello World!
  </div>
</DocumentFragment>
`;

exports[`renders without error 1`] = `<DocumentFragment />`;
